README
NAME
    Data::IEEE754::Tools - Various tools for understanding and manipulating
    the underlying IEEE-754 representation of floating point values

DESCRIPTION
    These tools give access to the underlying IEEE 754 floating-point 64bit
    representation used by many instances of Perl (see perlguts). They
    include functions for converting from the 64bit internal representation
    to a string that shows those bits (either as hexadecimal or binary) and
    back, functions for converting that encoded value into a more
    human-readable format to give insight into the meaning of the encoded
    values, and functions to manipulate the smallest possible change for a
    given floating-point value (which is the ULP
    <https://en.wikipedia.org/wiki/Unit_in_the_last_place> or "Unit in the
    Last Place").

  Compatibility
    Data::IEEE754::Tools works with 64bit floating-point representations.

    If you have a Perl setup which uses a larger representation (for
    example, "use Config; print $Config{nvsize}; # 16 => 128bit"), values
    reported by this module will be reduced in precision to fit the 64bit
    representation.

    If you have a Perl setup which uses a smaller representation (for
    example, "use Config; print $Config{nvsize}; # 4 => 32bit"), the
    installation will likely fail, because the unit tests were not set up
    for lower precision inputs. However, forcing the installation *might*
    still allow coercion from the smaller Perl NV into a true IEEE 754
    double (64bit) floating-point, but there is no guarantee it will work.

INSTALLATION
    To install this module, use your favorite CPAN client.

    For a manual install, type the following:

        perl Makefile.PL
        make
        make test
        make install

    (On Windows machines, you may need to use "dmake" instead of "make".)

AUTHOR
    Peter C. Jones "<petercj AT cpan DOT org>"

    Please report any bugs or feature requests emailing
    "<bug-Data-IEEE754-Tools AT rt.cpan.org>" or thru the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-IEEE754-Tools>, or
    thru the repository's interface at
    <https://github.com/pryrt/Data-IEEE754-Tools/issues>.

COPYRIGHT
    Copyright (C) 2016 Peter C. Jones

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See <http://dev.perl.org/licenses/> for more information.


CHANGES
Revision history for Perl module Data::IEEE754::Tools.

v0.019002 2018-Jun-17
    - META: add appveyor support and badges [uprev to disambiguate last fall's changes]

v0.019001 2017-Oct-03
    - Odd versions are development versions.  Features to be released
      in v0.020 were developed under the v0.019 revision
    - working on BUG #15: improved test suite on byteorder='1234' machine
        - t/01*.t fix binstr754_from_double() `like` test naming
        - t/01*.t switch to already-packed b64-strings, but didn't fix it
        - t/01*.t now uses the same as the module, if left(byteorder,4)='1234'
          (call it "cheating", but really, the forced-little-endian is nearly
          cheating as well; I have to assume that perl pack/unpack are 100%
          test coverage no matter how I test this)

v0.018004 2017-Oct-01
    - binary64_convertToHexString(): bugfix verified; now works on ivsize=4;
      remove debug code (#14)

v0.018003 2017-Sep-29
    - Completely reworked the function, so it deals with one nibble at a time,
      thus avoiding any overflow (easier to code than doing sub-32b groupings)
    - possible bug in t/01*.t, but unverified and unstudied

v0.018002 2017-Sep-25
    - Change from README.pod to README.md, because *.pod are built and put in
      the module documentation directory
    - Debugging roundoff on certain systems (not sure yet what the common
      aspect of the systems are, because bsd or mswin for the same perl version
      will both pass and fail -- my guess is 32bit vs 64bit machines, maybe
      in the integer processing, but I need data that will show me what's
      different, so I'm doing another "official" release that turns on debug
      printing during the test suite for known-buggy values.  (I would do alpha,
      but the smoke testers didn't run when I tried v0.017_xxx alpha versions,
      so sorry.  This release should be no more broken than v0.018 or v0.018001)

v0.018001 2017-Sep-24
    - for some reason, `make test` was invoking `make README.pod`: trying to fix

v0.018 2017-Jun-14
    - changing names per the naming-convention issue#6
      + make
        :floatingpoint, to_hex_floatingpoint(), to_dec_floatingpoint()
        aliases to
        :convertToString, convertToHexString(),
        convertToDecimalString()
      + make
        :raw754, hexstr754_from_double(), binstr754_from_double(),
        hexstr754_to_double(), binstr754_to_double()
        aliases to
        :internalString, convertToInternalHexString(),
        convertToInternalBinaryString(), convertFromInternalHexString(),
        convertFromInternalBinaryString()
      + for convertToDecimalString() and convertToHexString, add in optional
        conversionSpecification -- it specifiers the number of digits after the
        fractional-point.
    - meta: improve test coverage, documentation, development automation

v0.017
    - Odd versions are development versions.  Features to be released
      in v0.018 were developed under the v0.017 revision

v0.016 2016-Aug-29
    - had a bug in v0.014 Makefile which caused CPAN to not
      index; tried a couple of submissions, none of which
      really did what I wanted.  Will try to move up to next
      public release version number (v0.016), and ask PAUSE
      to forget the incorrect v0.14003.0 release, and hopefully
      re-index to this release.

v0.014 2016-Aug-29
    - feature request <https://rt.cpan.org/Ticket/Display.html?id=116155>
      + add many functions
        :ulp => ulp, nextUp, nextDown, nextAfter
        :info => isSignMinus, isNormal, isFinite, isNaN, isSignaling,
            isSignalingConvertedToQuiet, isCanonical, class, radix, totalOrder,
            totalOrderMag, compareFloatingValue, compareFloatingMag
        :signbit => copy, negate, abs, copySign, isSignMinus
    - feature request <https://rt.cpan.org/Ticket/Display.html?id=116154>
      + add :constants
    - feature request <https://rt.cpan.org/Ticket/Display.html?id=116153>
      + update ulp() with new, faster method

v0.013
    - Odd versions are development versions, used for developing and verifying
      new features and bug fixes.  There may have been one or more alpha
      subversions (v0.013_001, etc).

v0.012 2016-Jul-14
    - test coverage: Devel::Cover showed two conditionals which can only ever
        trigger one of the two (impossible to hit the opposite); remove the
        unnecessary conditional
    - bugfix <https://rt.cpan.org/Public/Bug/Display.html?id=116005>: get SNAN,
        expect QNAN
      - original code always expected QNAN, because all of the developer's
        machines automatically quiet SNAN to QNAN
      - v0.011_001: Test suite for :floatingpoint now converts SNAN to expect
        /[SQ]NAN/, because it doesn't really matter that much whether Perl
        silences a signaling NAN
    - bugfix <https://rt.cpan.org/Public/Bug/Display.html?id=116006>
      - getting NAN instead of expected value on nextup(NEGATIVE NORMAL) or
        nextdown(POSITIVE NORMAL)
      - v0.011_001: when LSB underflows (goes to -1), the hexification is too
          many F's; misinterpreted when the two were being recombined, so masked
          the MSB and LSB before merging
    - bugfix <https://rt.cpan.org/Public/Bug/Display.html?id=116004>
      - test suite getting -0.160000000000000003 when expecting -0.16
      - caused by systems with $Config{nvsize} > 8: code expected 64bit NV (some
        CPAN Testers have 128bit NV)
      - v0.011_002: change test suite to only compare enough significant figures
        for a 64bit double, so that even 128bit floats (quad?) will pass
      - v0.011_004: added documentation to explain that the module expects
        64bit NV, rounds off more-precise NV floats to 64bit, and will fail
        tests on 32bit NV (but might coerce to 64bit floats and might still

t/debug.pl
#!/bin/env perl
########################################################################
# Application Description
########################################################################
# Subversion Info
#   $Author: pryrtmx $
#   $Date: 2016-06-17 15:09:51 -0700 (Fri, 17 Jun 2016) $
#   $Revision: 169 $
#   $URL: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/t/debug.pl $
#   $Header: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/t/debug.pl 169 2016-06-17 22:09:51Z pryrtmx $
#   $Id: debug.pl 169 2016-06-17 22:09:51Z pryrtmx $
########################################################################
use warnings;
use strict;
use lib '../';
require 'Expand.pm';

use version 0.77; our $VERSION = version->declare('v1.000.' . sprintf '%03d', (q$Revision: 169 $ =~ /(\d+)/) );   # PAUSE compatible SVN-based version-object; can have extra

local $\ = "\n";

print $\ x 50;

print "$0: $VERSION";
print "\$Data::IEEE754::Expand::VERSION = ", $Data::IEEE754::Expand::VERSION;

print Data::IEEE754::Expand::get_raw_ieee754_hexstr_from_double(-.16);
print Data::IEEE754::Expand::get_raw_ieee754_binstr_from_double(-.16);
print '';

print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/binary/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/binary fraction/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/decimal/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/decimal fraction/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/hexadecimal/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/hexadecimal reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/hexadecimal fraction/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(-1.5, qw/hexadecimal fraction reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(0.16, qw/binary/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(0.16, qw/decimal/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(0.16, qw/hexadecimal/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(128.375, qw/binary reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(128.375, qw/binary fraction reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(128.375, qw/decimal reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(128.375, qw/decimal fraction reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(128.375, qw/hex reduce/);
print '';
print Data::IEEE754::Expand::ieee754_double2xxx_workhorse(0, qw/hex reduce/);
print '';

foreach my $h ( qw/0000000000000001 8000000000000001 7FF0000000000000 FFF0000000000000 7FF0000000000001 FFF0000000000001 7FF8000000000001 FFF8000000000001/ ) {
    my $v = Data::IEEE754::Expand::get_double_from_ieee754_hexstr( $h );
    printf "$h \t $v \t %23.16g\n", $v;
    print Data::IEEE754::Expand::ieee754_double2xxx_workhorse($v, qw/bin reduce/);
    print '';
}
exit;



t/rawstr.pl
#!/bin/env perl
########################################################################
# Verifies the following functions:
#   get_raw_ieee754_hexstr_from_double()
#   get_raw_ieee754_binstr_from_double()
#   get_double_from_ieee754_hexstr()
#   get_double_from_ieee754_binstr()
########################################################################
# Subversion Info
#   $Author: pryrtmx $
#   $Date: 2016-06-17 12:17:19 -0700 (Fri, 17 Jun 2016) $
#   $Revision: 168 $
#   $URL: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/t/rawstr.pl $
#   $Header: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/t/rawstr.pl 168 2016-06-17 19:17:19Z pryrtmx $
#   $Id: rawstr.pl 168 2016-06-17 19:17:19Z pryrtmx $
########################################################################
use warnings;
use strict;
use lib '../';
require 'Expand.pm';

use version 0.77; our $VERSION = version->declare('v1.000.' . sprintf '%03d', (q$Revision: 168 $ =~ /(\d+)/) );   # PAUSE compatible SVN-based version-object; can have extra

local $\ = "\n";

print $\ x 50;

print "$0: $VERSION";
print "\$Data::IEEE754::Expand::VERSION = ", $Data::IEEE754::Expand::VERSION;

my $v = -0.16;
print "rawhex754($v) =\t", Data::IEEE754::Expand::get_raw_ieee754_hexstr_from_double($v);
print "rawbin754($v) =\t", Data::IEEE754::Expand::get_raw_ieee754_binstr_from_double($v);
my $s = "BFC47AE147AE147B";
print "encode(0x$s) =\t", Data::IEEE754::Expand::get_double_from_ieee754_hexstr($s);
$s = "1011111111000100011110101110000101000111101011100001010001111011";
print "encode(0b$s) =\t", Data::IEEE754::Expand::get_double_from_ieee754_binstr($s);
print '';

$v = 1.0 + 1.0/16.0 + 5.0/256.0;
print "rawhex754($v) =\t", Data::IEEE754::Expand::get_raw_ieee754_hexstr_from_double($v);
print "rawbin754($v) =\t", Data::IEEE754::Expand::get_raw_ieee754_binstr_from_double($v);
$s = "3FF1500000000000";
print "encode(0x$s) =\t", Data::IEEE754::Expand::get_double_from_ieee754_hexstr($s);
$s = "0011111111110001010100000000000000000000000000000000000000000000";
print "encode(0b$s) =\t", Data::IEEE754::Expand::get_double_from_ieee754_binstr($s);
print '';
exit;



Makefile.PL
use ExtUtils::MakeMaker;
use 5.006;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
%mm_args = (
    'NAME'          => 'Data::IEEE754::Tools',
    'AUTHOR'        => 'Peter C. jones <petercj AT cpan.org>',
    'VERSION_FROM'  => 'lib/Data/IEEE754/Tools.pm',
    'ABSTRACT_FROM' => 'lib/Data/IEEE754/Tools.pm',
    'PREREQ_PM'     => {
        'Exporter'      => 5.57,    # needed for `use Exporter qw(import);` notation
        'version'       => 0.77,
        'warnings'      => 0,
        'strict'        => 0,
        'Carp'          => 0,
        'Config'        => 0,
    },
);

{
    no warnings;
    if( $ExtUtils::MakeMaker::VERSION >= '6.46' ) {
        $mm_args{META_MERGE} = {
            'meta-spec' => {
                version => '2',
                url     => 'http://search.cpan.org/perldoc?CPAN::Meta::Spec',
            },
            resources => {
                bugtracker => {
                    mailto  => 'bug-Data-IEEE754-Tools@rt.cpan.org',
                    web     => 'https://github.com/pryrt/Data-IEEE754-Tools/issues',
                },
                repository => {
                    url     => 'https://github.com/pryrt/Data-IEEE754-Tools.git',
                    web     => 'https://github.com/pryrt/Data-IEEE754-Tools',
                    type    => 'git',
                },
            },
            keywords => [ 'IEEE-754', 'floating point representation'],
        },
        # automatically determine the versions for 'provides'; MM->parse_version exists as of MM v6.31, so not a problem with >= 6.46
        #   => this comes out right when printed... but MYMETA.* shows values that look wrong
        $mm_args{META_MERGE}{provides} = {
            $mm_args{NAME} => {
                file => $mm_args{VERSION_FROM},
                version => MM->parse_version( $mm_args{VERSION_FROM} ),
            }
        };
    }
    if( $ExtUtils::MakeMaker::VERSION >= '6.31' ) {
        $mm_args{LICENSE} = 'perl_5';
    }
    if( $ExtUtils::MakeMaker::VERSION >= '6.48' ) {
        $mm_args{MIN_PERL_VERSION} = '5.6.0';
    }
    if( $ExtUtils::MakeMaker::VERSION >= '6.52' ) {
        $mm_args{CONFIGURE_REQUIRES} = {
            'ExtUtils::MakeMaker' => 0,
            'warnings' => 0,
            'strict' => 0,
            'Test::More' => 0.86,
            'constant' => 0,
            'Config' => 0,
        },
    }
    if( $ExtUtils::MakeMaker::VERSION >= '6.64' ) {
        $mm_args{TEST_REQUIRES} = {
            'warnings' => 0,
            'strict' => 0,
            'Test::More' => 0.86,
            'constant' => 0,
            'Config' => 0,
        };
    }
    if( $ExtUtils::MakeMaker::VERSION >= '6.18' ) {
        $mm_args{SIGN} = 1 if MM->can('signature_target') && $ENV{MM_SIGN_DIST};
    }
}

sub MY::postamble {
    my (undef,%h) = @_;
    #require Data::Dumper;
    #print STDERR Data::Dumper->Dump([\%h], [qw(mm_args{postamble})]);
    return <<'POSTAMBLE';
# TEST_D is used to test for the existence of a directory
TEST_D = $(ABSPERLRUN) -MExtUtils::Command -e test_d --

# run a test coverage report
testcover ::
	$(NOECHO) cover -test -report html

# run author-tests on the original
authtest ::
	$(TEST_D) xt && $(MAKE) test TEST_FILES='xt/*.t'

# run author-tests on the distribution
distauthtest :: disttest
	cd $(DISTVNAME) && $(TEST_D) xt && $(MAKE) test TEST_FILES='xt/*.t'

# addition to realclean: also delete coverage database, if testcover was run
realclean ::
	$(NOECHO) ( $(TEST_D) cover_db && $(RM_RF) cover_db ) || $(ECHO) realclean:: skip "rm -rf cover_db"
	$(NOECHO) ( $(TEST_F) MYMETA.json.lock && $(RM_RF) MYMETA.json.lock ) || $(ECHO) realclean:: skip "rm -rf MYMETA.json.lock"

# auto-generate the README from the lib/Data/IEEE754/Tools.pm
README.md :: lib/Data/IEEE754/Tools.pm
	podselect -section "NAME|SYNOPSIS|DESCRIPTION|COMPATIBILITY|INSTALLATION|AUTHOR|COPYRIGHT|LICENSE/!IEEE 754 Encoding" lib/Data/IEEE754/Tools.pm > README.pod
	pod2markdown README.pod README.md
	$(RM_F) README.pod

# auto-generate the LICENSE file from the lib/Data/IEEE754/Tools.pm
LICENSE :: lib/Data/IEEE754/Tools.pm
	podselect -section "AUTHOR|COPYRIGHT|LICENSE" lib/Data/IEEE754/Tools.pm > LICENSE.pod
	pod2text LICENSE.pod LICENSE
	$(RM_F) LICENSE.pod

docs :: README.md LICENSE

POSTAMBLE
}

WriteMakefile( %mm_args )


Expand.pm
#!/bin/env perl
########################################################################
=pod

=head1 NAME

Data::IEEE754::Expand - Takes an IEEE754 double, and expands it to SIGN * (1+FRACTION) * (2**EXPONENT)

=head1 SYNOPSIS

    use Data::IEEE754::Expand qw/expand_ieee754/;
    print expand_ieee754(-1.5, 'binary', 'fraction');

=head1 DESCRIPTION

Takes an IEEE754 double, and expands it to SIGN * (1+FRACTION) * (2**EXPONENT).

The L<IEEE754|https://en.wikipedia.org/wiki/IEEE_floating_point> standard describes
various floating-point encodings.  The double format (`binary64') is a 64-bit base-2
encoding, and correpsonds to the usual Perl floating value (NV). The format  includes the
sign (s), the power of 2 (q), and a coefficient (c): C<value = ((-1)**s) * (c) * (2**q)>.
The C<(-1)**s> term evaluates to the SIGN of the number, where s=0 means the sign
is +1 and s=1 means the sign is -1.
The coefficient is internally encoded as an implied 1 plus an encoded FRACTION,
which is itself encoded as a 52-bit integer divided by an implied 2**52.

L<Data::IEEE754>, or the equivalent L<perlfunc/pack> recipe L<dE<gt>>, does a
good job of converting a perl floating value (NV) into the big-endian bytes
that encode that value, but don't help you interpret.

This B<Data::IEEE754::Expand> module takes it to the next step, and provideds
the tools to expand the encoded floating point value into its component sign,
fraction, and exponent, so you can see how it is encoded internally.

=head2 FEATURES

=over

=item * will allow fraction to be represented as numerator/denominator or as a floating-point number

=item * will allow fraction to be represented in binary or hexadecimal or decimal

=item * will allow reducing the fraction (removing common factors of 2 or 16) to simplify interpretation

=back

=head1 ACKNOWLEDGEMENTS

=over

=item * L<What Every Compute Scientist Should Know About Floating-Point Arithmetic|https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html>

=item * L<Perlmonks: Integers sometimes turn into Reals after substraction|http://perlmonks.org/?node_id=1163025> for
inspiring me to go down the IEEE754-expansion trail in perl.

=item * L<Perlmonks: Exploring IEEE754 floating point bit patterns|http://perlmonks.org/?node_id=984141> as a resource
for how perl interacts with the various "edge cases" (+/-infinity, L<denormalized numbers|https://en.wikipedia.org/wiki/Denormal_number>,
signaling and quiet L<NaNs (Not-A-Number)|https://en.wikipedia.org/wiki/NaN>.

=item * L<Data::IEEE754>: I really wanted to use this module, but it didn't get me very far down the "Expand" track,
and included a lot of overhead modules for its install/test that I didn't want to require for B<Data::IEEE754::Expand>.
However, I was inspired by his byteorder-dependent anonymous subs (which were in turn derived from L<Data::MessagePack::PP>);
they were more efficient, on a per-call-to-subroutine basis, than my original inclusion of the if(byteorder) in every call to
the sub.

=back

=cut
########################################################################
# Subversion Info
#   $Author: pryrtmx $
#   $Date: 2016-06-17 15:09:51 -0700 (Fri, 17 Jun 2016) $
#   $Revision: 169 $
#   $URL: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/Expand.pm $
#   $Header: https://subversion.assembla.com/svn/pryrt/trunk/perl/Data-IEEE754-Expand/Expand.pm 169 2016-06-17 22:09:51Z pryrtmx $
#   $Id: Expand.pm 169 2016-06-17 22:09:51Z pryrtmx $
########################################################################
package Data::IEEE754::Expand;
use warnings;
use strict;
use Carp;
use Config;

use version 0.77; our $VERSION = version->declare('v0.002.' . sprintf '%03d', (q$Revision: 169 $ =~ /(\d+)/) );   # PAUSE compatible SVN-based version-object; can have extra

# Perl 5.10 introduced the ">" and "<" modifiers for pack which can be used to
# force a specific endianness.
if( $] lt '5.010' ) {
    my $str = join('', unpack("H*", pack 'L' => 0x12345678));
    if('78563412' eq $str) {        # little endian, so reverse byteorder
        *get_raw_ieee754_hexstr_from_double = sub { return uc unpack('H*' => reverse pack 'd'  => shift); };
        *get_raw_ieee754_binstr_from_double = sub { return uc unpack('B*' => reverse pack 'd'  => shift); };

        *get_double_from_ieee754_hexstr     = sub { return    unpack('d'  => reverse pack 'H*' => shift); };
        *get_double_from_ieee754_binstr     = sub { return    unpack('d'  => reverse pack 'B*' => shift); };
    } elsif('12345678' eq $str) {   # big endian, so keep default byteorder
        *get_raw_ieee754_hexstr_from_double = sub { return uc unpack('H*' =>         pack 'd'  => shift); };
        *get_raw_ieee754_binstr_from_double = sub { return uc unpack('B*' =>         pack 'd'  => shift); };

        *get_double_from_ieee754_hexstr     = sub { return    unpack('d'  =>         pack 'H*' => shift); };
        *get_double_from_ieee754_binstr     = sub { return    unpack('d'  =>         pack 'B*' => shift); };
    } else {
        # TODO: there are 4!-2 = 22 potential middle-endian encodings; there has to be a way to programmatically
        #   determine which it is, without listing all 22 elseif conditions.
        my %nibbles = ( 12 => undef, 34 => undef, 56 => undef, 78 => undef );
        foreach my $k (reverse sort keys %nibbles) {
            pos($str) = 0;
            $str =~ /$k/g;
            $nibbles{$k} = pos($str);
        }
            # OOPS: it's really 8!, because 64bit has 8 nibbles; need to generate a known unique ieee754.double, and
            # base the decoding on that;
        *get_raw_ieee754_hexstr_from_double = sub { undef };
        *get_raw_ieee754_binstr_from_double = sub { undef };

        *get_double_from_ieee754_hexstr     = sub { undef };
        *get_double_from_ieee754_binstr     = sub { undef };
    }
} else {
        *get_raw_ieee754_hexstr_from_double = sub { return uc unpack('H*' =>         pack 'd>' => shift); };
        *get_raw_ieee754_binstr_from_double = sub { return uc unpack('B*' =>         pack 'd>' => shift); };

        *get_double_from_ieee754_hexstr     = sub { return    unpack('d>' =>         pack 'H*' => shift); };
        *get_double_from_ieee754_binstr     = sub { return    unpack('d>' =>         pack 'B*' => shift); };
}

sub ieee754_double2xxx_workhorse($;@) {
    # $ = double precision ieee754 floating point: 64bit total = 1bit sign + 11bit exponent (offset) + 52bit fraction + 0bit implied 1+ before fraction
    #       Thus, value of ieee754 = ( (sign?-1:+1) * (implied_1 + fraction / (2**52)) * (2**(exponent-bias)) )
    # @ = arguments:
    #       'hex' overrides 'binary'
    #           'binary' [default] => 0b1.0101_0101_0101_0101_0101_0101_0101_0101_0101_0101_0101_0101_0101
    #           'hex' => 0x1.1234_5678_9abc_d   (13 nibbles = 13x4 bits = 52 fractional bits)
    #           'decimal' => only valid with 'fraction'; see below
    #       'fraction' overrides 'point'
    #           'point' [default] => examples above
    #           'fraction' =>
    #                 [     0b0_0100_0111_1010_1110_0001_0100_0111_1010_1110_0001_0100_0111_1011 ]
    #               + [ 1 + -------------------------------------------------------------------- ] * (2**-3)
    #                 [     0b1_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000 ]
    #           *** if 'fraction', then also allow 'decimal', so that you get { 1 + (decimal = 0 .. 2**52-1) / 2**52 } * sign * exponent
    #       'reduce' overrides 'allbits'
    #           'allbits' [default] => examples above
    #
    my $val = shift;
    my $rawbin = get_raw_ieee754_binstr_from_double($val);
    my ($sign, $exp, $fract) = ($rawbin =~ /(.)(.{11})(.{52})/);
    $sign += 0;                                         # convert to number
    $exp = oct("0b$exp") - 1023;                        # convert to number, and adjust for the IEEE754 Exponent Bias for a double (binary64)
    my ($num, $den, $out) = ($fract, '0'x52, '');

    if( $exp == 1024 ) {
        #INF or NaN
        print STDERR "\nINF or NaN: $sign $exp $fract\n";
        return "".$val; # stringify and return
    } elsif ( $exp == -1023 ) {
        #Zero or Denormal
        print STDERR "\nZERO or Denormal: $sign $exp $fract\n";
        return "0" if $fract eq '0000000000000000000000000000000000000000000000000000';
    }

    # parse options
    my ($base, $drawFraction, $reduce) = ('',0,0);
    foreach(@_) {
        $base .= 'x'        if /^hex(|adecimal)$/i;
        $base .= 'b'        if /^bin(|ary)$/i;
        $base .= 'd'        if /^dec(|imal)$/i;
        $drawFraction |= 1  if /^fraction$/i;
        $reduce |= 1        if /^reduce$/i;
    }
    $base = 'b'     if '' eq $base; # default
    unless(1 == length($base)) {
        my %replace = (b => 'binary', d => 'decimal', h => 'hexadecimal');
        my $msg = "Bases " . join(" and ", map { $replace{lc $_} } split(//, $base)) . " are mutuall exclusive: chose";
        $base = ($base =~ /b/i) ? 'b' : ($base =~ /h/i) ? 'h' : 'd';    # priority = binary > hexadecimal > decimal (so if all three, binary chosen; if two, then prefer binary over hex, and hex over decimal)
        carp "$msg $base";
    }

    my $fval = (abs($val) / (2**$exp) - 1) * (2**52);

    # if hex then convert 52bit fract to 13nibble fract
    if($base eq 'x') {
        $den = '0000000000000'; # implied 1 before... but don't want that there, yet

        # I'd like to just the integer form of $fval, converted to hex
        #   however, on machines with $Config{ivsize}=4, it cannot convert to a 13nibble hex (because a 4byte value is only 8 nibbles of integer space)
        if($Config{ivsize}>4) {
            $num = substr sprintf('%013X', $fval), -13;
        } else {
            my ($left, $right) = ( $fract =~ /^(\d{20})(\d{32})$/ );
            $num = sprintf "%05X%08X", oct("0b$left"), oct("0b$right");
        }
    }

    if($base eq 'd') {                                              # decimal: convert to num & den, optionally reduce, then add commas every 3 characters
        if($drawFraction) {                                             # 1,234,567
            my $p = 52;
            if($reduce) {
                while(0 == $fval % 2) {
                    --$p;
                    $fval /= 2;
                }
            }
            $num = sprintf '%.0f', abs($fval);
            $den = sprintf '%.0f', (2**$p);
            foreach ($num, $den) { s/(\d{1,3}?)(?=(\d{3})+$)/$1,/g; }   # http://www.perlmonks.org/?node_id=117697: the fewest number of digits that has only multiples of 3 digits beyond it
        } else {                                                        # .123,456,7
            $num = sprintf "%.24f", abs($val) / (2**$exp);
            if($reduce) {
                $num =~ s/0+$//g;   # remove all trailing zeroes
            }
            my ($in,$fr) = split /\./, $num;
            foreach ($in) { s/(\d{1,3}?)(?=(\d{3})+$)/$1,/g; }          # integer needs 1,234,567-style
            foreach ($fr) { s/(\d{3})(?=\d)/$1_/g; }                    # 3 digits, followed by a non-capturing match of at least 1 digit
            $out = ( $sign ? '-' : '+' ) . "${in}.${fr} * (2**$exp)";
        }
    } elsif( $base =~ /^[xb]$/ ) {                                      # hex or binary: optionally reduce, add _ every four chars, and for fractional, prefix num and denom with 0b or 0x
        if($reduce) {
            $num =~ s/0+$//g;    # remove all trailing zeroes
            $den = '0' x length($num);
        }
        foreach ($num, $den) { s/([[:xdigit:]]{1,4})/_$1/g; }
        if($drawFraction) { # fractional
            $num = "0${base}0$num";
            $den = "0${base}1$den";
        } else {
            ($out = ( $sign ? '-' : '+' ) . "0${base}1.$num * (2**$exp)") =~ s/\._/./;
        }

    }

    # convert to fraction
    if($drawFraction) {
        my $dash = '-' x length($den);
        $num  = (' ' x length($den)) . $num;
        $num  = substr $num, -length($den);
        $out     = "";
        $out    .= "  [     $num ]\n";
        $out    .= '' . ( $sign ? '-' : '+' ) . ' [ 1 + ' . $dash . ' ] * (2**' . $exp . ") = $val\n";
        $out    .= "  [     $den ]";
    }

    if(1) { # debug only
        local $" = ", ";
        $out = "ieee754_double2xxx_workhorse(@{[$val, @_]})\n\n$out";
        $out .= "\n\n-----\n\n";
    }

    return $out;

    # 3.TODO = add callers that automate the options
    # 2.TODO (v0.004) = rename:
    #       module = Data::IEEE754::Tools
    #       raw functions = hexstr754_from_double() and hexstr754_to_double() (or binstr*)
    #       workhorse = expand_ieee754_workhorse()
    #       callers = expand_to_*(NV)
    # 1.TODO (v0.003) = since I now have the reverse functions, I have an idea for how to rework this, to get rid of some of the conversions
    #   between hex/binary/decimal
    #       x. use binstr764_from_double() as currently handled
    #       x. check exponent
    #           * ==emax+1 => isInf or isNaN, so just stringify: return $out = "".$in
    #           * ==emin-1 => isDenormal or isZero; if isDenormal, exp=emin; if isZero, exp=0; continue
    #           * else     => normal; continue
    #       c. determine whether it's hexadecimal/decimal/binary output
    #           * binary: I've got everything I need: numer comes from binary $fract; continue
    #           * hex: numer comes from hexstr754_from_double():
    #               new $fract = substr $hex, -13 (rightmost 13 chars)
    #               if isDenormal, 0+, else 1+
    #               continue
    #           * decimal:
    #               if drawFraction
    #                   (left,right) = ( fract =~ /(.{20})(.{32})$/ );  # split into 20bit and 32bit numbers, which both fit within 4byte integer
    #                   numer = (2.0**32)*oct("0b$left") + 1.0*oct("0b$right") -- this keeps it floating-point, so it will fit
    #                   denom = (2.0**52)
    #               else
    #                   numer = $fval                       # might be wrong for denormal; will need to look into it
    #                   out = sign . numer . "2**$exp"
    #       d. reduce (if applicable)
    #           * bin or hex: remove trailing 0s
    #           * decimal: if drawFraction, divide-by-2**k; else remove trailing 0s
    #       e. add commas/underscores (per existing code)
    #       f. output
}

1;
